performance.md

```diff
@@ -0,0 +1,341 @@
+important".
+
+The most significant qualities might be `safety` and `speed`. 
+
+By default, Lisp considers code safety to be much more important than
+speed. But you may adjust the weight for more aggressive optimization.
+
+~~~lisp
+* (defun max-original (a b)
+    (max a b))
+MAX-ORIGINAL
+
+* (disassemble 'max-original)
+; disassembly for MAX-ORIGINAL
+; Size: 144 bytes. Origin: #x52D450EF
+...
+
+* (defun max-with-speed-3 (a b)
+    (declare (optimize (speed 3) (safety 0)))
+    (max a b))
+MAX-WITH-SPEED-3
+
+* (disassemble 'max-with-speed-3)
+; disassembly for MAX-WITH-SPEED-3
+; Size: 92 bytes. Origin: #x52D452C3
+...
+~~~
+
+As you can see, the generated assembly code is much shorter, which means that
+unused code are pruned.
+
+### Type Hints
+
+As mentioned in the [*Type System*][sec:type] chapter, Lisp has a very
+powerful type system. You may provide type hints so that the compiler may
+reduce size of generated code.
+
+~~~lisp
+* (defun max-with-type (a b)
+    (declare (optimize (speed 3) (safety 0)))
+    (declare (type integer a b))
+    (max a b))
+MAX-WITH-TYPE
+
+* (disassemble 'max-with-type)
+; disassembly for MAX-WITH-TYPE
+; Size: 42 bytes. Origin: #x52D48A23
+~~~
+
+The size of generated assembly code shrunk to about 1/3 of the size. What
+about speed?
+
+~~~lisp
+* (time (dotimes (i 10000) (max-original 100 200)))
+Evaluation took:
+  0.000 seconds of real time
+  0.000107 seconds of total run time (0.000088 user, 0.000019 system)
+  100.00% CPU
+  361,088 processor cycles
+  0 bytes consed
+
+* (time (dotimes (i 10000) (max-with-type 100 200)))
+Evaluation took:
+  0.000 seconds of real time
+  0.000044 seconds of total run time (0.000036 user, 0.000008 system)
+  100.00% CPU
+  146,960 processor cycles
+  0 bytes consed
+~~~
+
+You see, by specifying type hints, our code runs much faster!
+
+But wait...What happens if we declare wrong types? The answer is: it depends.
+
+For example, SBCL treats type declarations in a [special way][sbcl-type]. It
+performs different levels of type checking according to the safety level. If
+safety level is set to 0, no type checking will be performed. Thus wrong type
+specifier might bring a lot of damage.
+
+### More on Type Declaration
+
+If you try to evaluate `declare` form in the top level, you might get the
+following error:
+
+~~~lisp
+Execution of a form compiled with errors.
+Form:
+  (DECLARE (SPEED 3))
+Compile-time error:
+  There is no function named DECLARE.  References to DECLARE in some contexts
+(like starts of blocks) are unevaluated expressions, but here the expression is
+being evaluated, which invokes undefined behaviour.
+   [Condition of type SB-INT:COMPILED-PROGRAM-ERROR]
+~~~
+
+This is because type declarations have [scopes][declare-scope]. In the
+examples above, we have seen type declarations applied to a function. 
+
+During development it is usually useful to raise importance of safety in order
```

Comment:
raise "the" importance of safety "," in order

Reply:
ACK.